### PR TITLE
Implement `in_place_scope`

### PR DESCRIPTION
**Pull request by rocallahan**
As discussed in #562.

### CODE REVIEW

**Comment by rocallahan** [comment]
I fixed those check/format issues and repushed.

File: rayon-core/src/scope/mod.rs [hunk 1]
```diff
@@ -381,17 +394,70 @@ where
...
+///
+/// If a panic occurs, either in the closure given to `scope()` or in
+/// any of the spawned jobs, that panic will be propagated and the
+/// call to `scope()` will panic. If multiple panics occurs, it is
+/// non-deterministic which of their panic values will propagate.
+/// Regardless, once a task is spawned using `scope.spawn()`, it will
+/// execute, even if the spawning task should later panic. `scope()`
+/// returns once all spawned jobs have completed, and any panics are
+/// propagated at that point.
+pub fn external_scope<'scope, OP, R>(op: OP) -> R
```
Comment (nikomatsakis) [inline rayon-core/src/scope/mod.rs:422]: I don't love the name 'external scope' -- it may not be external, after all, if this is running on a worker thread, right?

Maybe `in_place_scope` or something like that?

**Review by nikomatsakis — COMMENTED** [review]
One quick nit, didn't read the code in detail.

File: rayon-core/src/scope/mod.rs (hunk 1 above)
Comment (rocallahan) [reply rayon-core/src/scope/mod.rs:422]: `in_place_scope` sounds fine to me but @cuviper should weigh in I guess.

Comment (cuviper) [reply rayon-core/src/scope/mod.rs:422]: `direct_scope`? `immediate_scope`? `local_scope`?

I'm not sure what will be most intuitive, that users might naturally guess what it means compared to `scope` alone.

(My own prior experimentation called it `raw_scope`, but I think that misses the mark.)

Comment (rocallahan) [reply rayon-core/src/scope/mod.rs:422]: `this_thread_scope`?

File: rayon-core/src/latch.rs [hunk 2]
```diff
@@ -218,6 +218,7 @@ impl<'r> Latch for SpinLatch<'r> {
 
 /// A Latch starts as false and eventually becomes true. You can block
 /// until it becomes true.
+#[derive(Debug)]
```
Comment (lu-zero) [inline rayon-core/src/latch.rs:221]: might be nice as a separate commit.

File: rayon-core/src/scope/mod.rs [hunk 3]
```diff
@@ -381,17 +394,70 @@ where
...
-            base: ScopeBase::new(owner_thread),
+/// Creates a "fork-join" scope `s` and invokes the closure with a
+/// reference to `s`. This closure can then spawn asynchronous tasks
+/// into `s`. Those tasks may run asynchronously with respect to the
+/// closure; they may themselves spawn additional tasks into `s`. When
+/// the closure returns, it will block until all tasks that have been
+/// spawned into `s` complete.
+///
+/// This is just like `scope()` except the closure runs on the same thread
+/// that calls `external_scope()`. Only work that it spawns runs in the
```
Comment (lu-zero) [inline rayon-core/src/scope/mod.rs:409]: Only calls to spawn()? Would be good to have the fifo variant at the same time?

File: rayon-core/src/scope/mod.rs (hunk 1 above)
Comment (lu-zero) [reply rayon-core/src/scope/mod.rs:422]: `in_place_scope` or `immediate_scope` sound both good. The test is good to explain the use-case.

Comment (rocallahan) [reply rayon-core/src/scope/mod.rs:422]: @cuviper can you sign off on `in_place_scope`?

Comment (cuviper) [reply rayon-core/src/scope/mod.rs:422]: Yeah, that's fine.

File: rayon-core/src/scope/mod.rs (hunk 3 above)
Comment (rocallahan) [reply rayon-core/src/scope/mod.rs:409]: How about we get this reviewed and landed first so I don't duplicate any mistakes I've made?

**Comment by rocallahan** [comment]
I've uploaded new commits with the #[derive(Debug)] split out and the name changed to `in_place_scope`.

File: rayon-core/src/scope/mod.rs (hunk 3 above)
Comment (lu-zero) [reply rayon-core/src/scope/mod.rs]: Sounds good to me.

File: rayon-core/src/latch.rs [hunk 4]
```diff
@@ -329,6 +327,41 @@ impl AsCoreLatch for CountLatch {
     }
 }
 
+#[derive(Debug)]
```
Comment (nikomatsakis) [inline rayon-core/src/latch.rs:330]: Can you document the role of a "count lock latch", at least briefly?

File: rayon-core/src/scope/mod.rs [hunk 5]
```diff
@@ -37,21 +38,31 @@ pub struct ScopeFifo<'scope> {
...
-    /// may be executing on different worker threads, though they
-    /// should always be within the same pool of threads)
-    owner_thread_index: usize,
+enum ScopeLatch {
+    /// A latch for scopes created on a rayon thread which will participate in work-
+    /// stealing while it waits for completion. This thread is not necessarily part
+    /// of the same registry as the scope itself!
+    Stealing {
+        latch: CountLatch,
+        registry: Arc<Registry>,
```
Comment (nikomatsakis) [inline rayon-core/src/scope/mod.rs:47]: I'm curious what @cuviper thinks: I'd be inclined not to store this in a field but instead to pass an `&Arc<Registry>` as an argument to `ScopeLatch::set`. It avoids an extra atomic increment on every `scope`.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: rayon-core/src/scope/mod.rs (hunk 5 above)
Comment (rocallahan) [reply rayon-core/src/scope/mod.rs:47]: What about the cross-registry case? If we call `ThreadPool::in_place_scope()` on registry B from a worker thread of registry A, when a job completes on a worker thread of registry B, we would need to call `ScopeLatch::set` with a reference to registry A but we won't be able to find one.

Comment (nikomatsakis) [reply rayon-core/src/scope/mod.rs:47]: I must not have my head in the right space; I don't quite follow. Maybe you can add a nice comment into the code detailing how that scenario plays out and why this field is necessary?

Comment (rocallahan) [reply rayon-core/src/scope/mod.rs:52]: I added a comment but I'm not sure I've made it any clearer than what I already said.

When a job completes in registry B we only have a reference to registry B, but we may need to wake the thread in registry A, for which we need a reference to registry A.

Comment (nikomatsakis) [reply rayon-core/src/scope/mod.rs:52]: Thanks. That helps. I'll try to make some time to read it over. I'm still contemplating what happened in the old code for this logic -- I guess the point is that this case didn't arise, because we always executed the scope's closure over within the target registry, so there was no cross-registry case? (Or at least, it was handled by the logic that moved over into the other registry?)

Comment (cuviper) [reply rayon-core/src/scope/mod.rs:52]: > I guess the point is that this case didn't arise, because we always executed the scope's closure over within the target registry, so there was no cross-registry case? (Or at least, it was handled by the logic that moved over into the other registry?)

Correct -- it would have ended up calling `scope` -> `in_worker` -> `in_worker_cross` to handle that case.

**Comment by nikomatsakis** [comment]
I'm happy. I would be willing to merge.

File: rayon-core/src/scope/mod.rs [hunk 6]
```diff
@@ -511,12 +584,20 @@ impl<'scope> ScopeFifo<'scope> {
...
-            registry: owner_thread.registry().clone(),
+            registry: registry,
             panic: AtomicPtr::new(ptr::null_mut()),
-            job_completed_latch: CountLatch::new(),
+            job_completed_latch: ScopeLatch::new(owner),
+            marker: PhantomData,
+        }
+    }
+
+    fn new_in_place(registry: Arc<Registry>) -> Self {
```
Comment (cuviper) [inline rayon-core/src/scope/mod.rs:596]: I think this would be better named `new_blocking`, since in-place sometimes uses the stealing `new` too.

File: rayon-core/src/lib.rs [hunk 7]
```diff
@@ -78,7 +78,7 @@ mod test;
 
 pub use self::join::{join, join_context};
 pub use self::registry::ThreadBuilder;
-pub use self::scope::{scope, Scope};
+pub use self::scope::{in_place_scope, scope, Scope};
```
Comment (cuviper) [inline rayon-core/src/lib.rs:81]: We should re-export this from the `rayon` crate too.

**Review by cuviper — COMMENTED** [review]
Looks great, just a couple small things...

File: rayon-core/src/scope/mod.rs (hunk 6 above)
Comment (rocallahan) [reply rayon-core/src/scope/mod.rs]: Done

File: rayon-core/src/lib.rs (hunk 7 above)
Comment (rocallahan) [reply rayon-core/src/lib.rs:81]: Done
